app/models/service/resource_linking.rb

```diff
@@ -0,0 +1,36 @@
+module Service::ResourceLinking
+  extend ActiveSupport::Concern
+
+  def add_provider_vms(options)
+    # required keys in options hash:
+    #   :target_id, the id of the provider to refresh
+    #   :array_of_uid_ems, an array of VMs' uid_ems
+    #
+    options = {
+      :target_class => "ExtManagementSystem",
+      :name         => "Linking VMs to service #{name} ID: [#{id}]",
+      :userid       => evm_owner.userid,
+      :sync_key     => guid,
+      :service_id   => id,
+      :zone         => my_zone
+    }.merge(normalize_inputs!(options))
+
+    _log.info("NAME [#{options[:name]}] for user #{evm_owner.userid}")
+
+    job = Service::LinkingWorkflow.create_job(options)
+    return job
+  rescue => err
+    _log.log_backtrace(err)
+    raise
+  end
+
+  def normalize_inputs!(options)
+    raise _("no target_id defined for linking targets to service") unless options[:target_id]
+
+    array_of_uid_ems = options[:array_of_uid_ems]
+    options[:array_of_uid_ems] = array_of_uid_ems.kind_of?(Array) ? array_of_uid_ems.compact.uniq : [array_of_uid_ems]
```

Comment:
These two lines can be simplified:
`options[:array_of_uid_ems] = Array(options[:array_of_uid_ems]).compact.uniq`